AndroidP/app/src/main/java/com/novoda/androidp/pt651_display_cutout/CutoutActivity.kt

```diff
@@ -0,0 +1,97 @@
+package com.novoda.androidp.pt651_display_cutout
+
+import android.content.SharedPreferences
+import android.os.Bundle
+import android.preference.PreferenceManager
+import android.support.v7.app.AppCompatActivity
+import android.view.View
+import android.view.WindowManager
+import android.widget.RadioGroup
+import com.novoda.androidp.R
+
+class CutoutActivity : AppCompatActivity() {
+
+    private lateinit var linearLayout: View
+    private val CUTOUT_MODE = "cutout_mode"
+    private val CUTOUT_USE_INSET = "cutout_use_inset"
+    private val CUTOUT_MODE_DEFAULT = WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_DEFAULT
+
+    val cutoutModeMap = mapOf(
+            R.id.rd_cutout_default to WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_DEFAULT,
+            R.id.rd_cutout_always to WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_ALWAYS,
+            R.id.rd_cutout_never to WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_NEVER)
+
+    val cutoutUseInsetMap = mapOf(
+            R.id.rd_cutout_yes to true,
+            R.id.rd_cutout_no to false
+    )
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_cutout)
+        linearLayout = findViewById(R.id.main_layout)
+
+        val pref = PreferenceManager.getDefaultSharedPreferences(this)
+        setLayoutCutoutMode(pref)
+        setupTitleView(pref)
+        setupRadioButtonForCutoutMode(pref)
+        setupRadioButtonForInset(pref)
+
+    }
+
+    private fun setLayoutCutoutMode(pref: SharedPreferences) {
+        window.attributes.layoutInDisplayCutoutMode = cutoutModeMap[pref.getInt(CUTOUT_MODE, CUTOUT_MODE_DEFAULT)] ?: CUTOUT_MODE_DEFAULT
+    }
+
+    private fun setupTitleView(pref: SharedPreferences) {
+        val titleView = findViewById<View>(R.id.tv_cutout_title)
+        titleView.setOnApplyWindowInsetsListener { view, windowInsets ->
+            windowInsets.displayCutout?.let {
+                if (isUseInsetSelected(pref)) {
+                    val safeInsetTop = windowInsets.displayCutout.safeInsetTop
+                    titleView.y = titleView.top + safeInsetTop.toFloat()
+                }
+            }
+
+            windowInsets
+        }
+    }
+
+    private fun isUseInsetSelected(pref: SharedPreferences) = cutoutUseInsetMap[pref.getInt(CUTOUT_USE_INSET, 0)] ?: false
+
+    private fun setupRadioButtonForCutoutMode(pref: SharedPreferences) {
+        val radioGroup = findViewById<RadioGroup>(R.id.rd_cutout_mode)
+        radioGroup.check(pref.getInt(CUTOUT_MODE, R.id.rd_cutout_default))
+        radioGroup.setOnCheckedChangeListener { radioGroup, checkedId ->
+            pref.edit().putInt(CUTOUT_MODE, checkedId).commit()
```

Comment:
Why not simply adding an intent extra when restarting and reading it in `onCreate`? Seems less of a roundtrip

Reply:
I want to persist it so I can completely kill the app and start it in the selected mode I want. This way I know it's working fine from scratch

Reply:
You could simply use [`persistableMode`](https://developer.android.com/reference/android/R.attr.html#persistableMode) then

Reply:
Oh wow that's powerful, I didn't know about it. Let me try it 💯

Reply:
Ok it's not that straight forward because I'm re-launching the same activity.
The problem I'm facing is `onSaveInstanceState` on the first instance of the activity is only called _after_ the second one is launched, so I don't have a `persistentState` to read stuff from.

If I make the activity `singleTask` so there's only one instance then `onSaveInstanceState` is never called because the instance is never destroyed.

Overall I see too much hassle for something that's not the important thing to showcase here :/

Reply:
Yeah becoming way too much hassle — to me it's a big YAGNI to save the state in the preferences in the first place anyway